lib/Transforms/IPO/StructFieldCacheAnalysis.cpp

```diff
@@ -36,30 +46,672 @@ class StructFieldCacheAnalysisPass : public ModulePass {
+  outs() << "Stats:\n";
+  for (auto &it : StructFieldAccessInfoMap){
+    for (auto i = 0; i < DebugStats::DS_MaxNumStats; i++){
+      StatCounts[i] += it.second->getStats(i);
+      if (i == DebugStats::DS_GepUnknownUse){
+        it.second->printUnknownOpcodes(outs());
+      }
+    }
+  }
+  for (auto i = 0; i < DebugStats::DS_MaxNumStats; i++){
+    if (StatCounts[i]){
+      outs() << "Case " << StatNames[i] << " was found " << StatCounts[i] <<  " times\n";
+    }
+  };
+  FILE_OS << "GEP as Arg," << StatCounts[DebugStats::DS_GepPassedIntoFunc] << "\n";
+  outs().changeColor(raw_ostream::BLUE);
+  outs() << "Stats are stored into " << "/tmp/SFCA-"+CurrentModule.getName().str()+".csv" << "\n";
+  outs().resetColor();
+  outs() << "----------------------------------------------------------------- \n";
+  FILE_OS.close();
+}
+
+static void performIRAnalysis(Module &M,
+                              StructFieldAccessManager* StructManager)
+{
+  // Find all global structs
+  for (auto &G : M.globals()){
+    // Only process globals defined in current module (the scope of whole program)
+    if (G.isDeclaration())
+      continue;
+    DEBUG_WITH_TYPE(DEBUG_TYPE_IR, dbgs().changeColor(raw_ostream::YELLOW));
+    // G is always a pointer
+    if (G.getValueType()->isStructTy()){
+      DEBUG_WITH_TYPE(DEBUG_TYPE_IR, dbgs() << "Found a global defined as struct: " << G << "\n");
+      auto* structInfoPtr = StructManager->createOrGetStructFieldAccessInfo(G.getValueType(), StructFieldAccessManager::StructDefinitionType::SDT_GlobalStruct);
+      assert(structInfoPtr);
+      structInfoPtr->analyzeUsersOfStructValue(&G);
+    }
+    // Case for struct*
+    else if (G.getValueType()->isPointerTy() && G.getValueType()->getPointerElementType()->isStructTy()){
+      DEBUG_WITH_TYPE(DEBUG_TYPE_IR, dbgs() << "Found a global has struct* type: " << G << "\n");
+      auto* structInfoPtr = StructManager->createOrGetStructFieldAccessInfo(G.getValueType()->getPointerElementType(), StructFieldAccessManager::StructDefinitionType::SDT_GlobalStructPtr);
+      assert(structInfoPtr);
+      structInfoPtr->analyzeUsersOfStructPointerValue(&G);
+    }
+    // Case for struct**
+    else if (G.getType()->isPointerTy() && G.getType()->getPointerElementType()->isPointerTy() && G.getType()->getPointerElementType()->getPointerElementType()->isStructTy()){
+      DEBUG_WITH_TYPE(DEBUG_TYPE_IR, dbgs() << "Found a global has struct** type: " << G << " but we ignored this\n");
+      StructManager->addStats(StructFieldAccessManager::DebugStats::DS_StructPtrPtr);
+    }
+    DEBUG_WITH_TYPE(DEBUG_TYPE_IR, dbgs().resetColor());
+  }
+
+  // Find all structs declared by allocas
+  for (auto &F : M){
+    if (F.isDeclaration())
+      continue;
+    //Find all alloca of structs inside the function body
+    for (auto &BB : F){
+      for (auto &I: BB){
+        if (I.getOpcode() == Instruction::Alloca){
+          assert(I.getType()->isPointerTy());
```

Comment:
Again, it was the time when I am not sure about some of the behaviors... Thank you for catching it!